test-suite-custom-generators/src/main/java/io/micronaut/sourcegen/custom/visitor/innerTypes/GenerateInnerTypeInEnumVisitor.java

```diff
@@ -0,0 +1,153 @@
+import io.micronaut.inject.ast.ClassElement;
+import io.micronaut.inject.processing.ProcessingException;
+import io.micronaut.inject.visitor.TypeElementVisitor;
+import io.micronaut.inject.visitor.VisitorContext;
+import io.micronaut.sourcegen.annotations.Builder;
+import io.micronaut.sourcegen.custom.example.GenerateInnerTypes;
+import io.micronaut.sourcegen.generator.SourceGenerator;
+import io.micronaut.sourcegen.generator.SourceGenerators;
+import io.micronaut.sourcegen.model.ClassDef;
+import io.micronaut.sourcegen.model.EnumDef;
+import io.micronaut.sourcegen.model.FieldDef;
+import io.micronaut.sourcegen.model.InterfaceDef;
+import io.micronaut.sourcegen.model.MethodDef;
+import io.micronaut.sourcegen.model.PropertyDef;
+import io.micronaut.sourcegen.model.RecordDef;
+import io.micronaut.sourcegen.model.TypeDef;
+
+import javax.lang.model.element.Modifier;
+import java.util.List;
+
+@Internal
+public final class GenerateInnerTypeInEnumVisitor implements TypeElementVisitor<GenerateInnerTypes, Object> {
+
+    @Override
+    public @NonNull VisitorKind getVisitorKind() {
+        return VisitorKind.ISOLATING;
+    }
+
+    @Override
+    public void visitClass(ClassElement element, VisitorContext context) {
+        SourceGenerator sourceGenerator = SourceGenerators.findByLanguage(context.getLanguage()).orElse(null);
+        if (sourceGenerator == null) {
+            return;
+        }
+
+        EnumDef enumDef = getEnumDef(element, context.getLanguage());
+
+        context.visitGeneratedSourceFile(enumDef.getPackageName(), enumDef.getSimpleName(), element)
+            .ifPresent(generatedFile -> {
+                try {
+                    generatedFile.write(writer -> sourceGenerator.write(enumDef, writer));
+                } catch (Exception e) {
+                    throw new ProcessingException(element, e.getMessage(), e);
+                }
+            });
+    }
+
+    private static EnumDef getEnumDef(ClassElement element, VisitorContext.Language language) {
+        String enumClassName = element.getPackageName() + ".MyEnumWithInnerTypes";
+
+        EnumDef innerEnum = getInnerEnumDef();
+
+        RecordDef innerRecord = getInnerRecordDef();
+
+        ClassDef innerClass = getInnerClassDef(language);
+
+        InterfaceDef innerInterface = getInnerInterfaceDef();
+
+        //outer enum
+        return EnumDef.builder(enumClassName)
+            .addModifiers(Modifier.PUBLIC)
+            .addEnumConstant("A")
+            .addEnumConstant("B")
+            .addEnumConstant("C")
+            .addInnerType(innerEnum)
+            .addInnerType(innerRecord)
+            .addInnerType(innerClass)
+            .addInnerType(innerInterface)
+            .addMethod(MethodDef.builder("myName")
+                .addModifiers(Modifier.PUBLIC)
+                .returns(String.class)
+                .build((aThis, parameters) ->
+                        aThis.invoke("toString", TypeDef.STRING, List.of()).returning()))
```

Comment:
Why do you keep using `List.of`? There is an array variation, and you can omit the parameters.